src/main/resources/hudson/tasks/junit/History/index.jelly

```diff
@@ -24,84 +24,136 @@ THE SOFTWARE.
 
 <!-- Displays the chart that show how long builds are taking -->
 <?jelly escape-by-default='true'?>
-<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:l="/lib/layout" xmlns:bs="/bootstrap5" xmlns:c="/charts">
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:l="/lib/layout" xmlns:bs="/bootstrap5" xmlns:c="/charts" xmlns:f="/lib/form" xmlns:dd="/lib/layout/dropdowns">
   <bs:layout title="${%title(it.testObject.displayName)}">
+    <j:set var="count" value="${it.asInt(request.getParameter('count'),100)}"/>
     <j:set var="start" value="${it.asInt(request.getParameter('start'),0)}"/>
-    <j:set var="end" value="${it.asInt(request.getParameter('end'),start+24)}"/>
+    <j:set var="end" value="${it.asInt(request.getParameter('end'),start+count-1)}"/>
+    <j:set var="historySummary" value="${it.retrieveHistorySummary(start, end)}"/>
+    <j:set var="trendChartJson" value="${historySummary.trendChartJson}"/>
+    <j:set var="trendChartJsonStr" value="${historySummary.trendChartJsonStr}"/>
+    <j:set var="buildMap" value="${historySummary.buildMapJson}"/>
     <st:adjunct includes="io.jenkins.plugins.echarts"/>
     <st:adjunct includes="io.jenkins.plugins.bootstrap5"/>
     <st:adjunct includes="io.jenkins.plugins.echarts-trend-default-setup"/>
     <link rel="stylesheet" href="${resURL}/plugin/junit/history/history.css"/>
 
     <st:once>
       <c:chart-setup id="test-history"/>
-      <script>var view =
-        <st:bind value="${it}"/>
+      <script>
+      var view = <st:bind value="${it}"/>
+      var start = ${start}
+      var end = ${end}
+      var count = ${count}
+      var trendChartJson = ${trendChartJson}
+      var trendChartJsonStr = ${trendChartJsonStr}
+      var rootUrl = "${app.rootUrl}"
+
+      function onBuildWindowChange(selectObj) {
+          var idx = selectObj.selectedIndex;
+          var count = selectObj.options[idx].value;
+          document.location = "${app.rootUrl}${it.testObject.url}/history?start=${start}&amp;count=" + count
+      }
```

Comment:
we shouldn't really use inline JavaScript (due to ongoing work to introduce a Content-Security-Policy)

see the guidance at: https://www.jenkins.io/doc/developer/security/csp/

there's a way to avoid the `st:bind` trick but I can't remember it off the top of my head (cc @daniel-beck)

unsure if blocker

Reply:
> a way to avoid the `st:bind` trick

I think you are referring to `data-*` attributes?

Reply:
Yes in some way

Reply:
Not data-* specifically but binding the variable to a window variable like in https://github.com/jenkinsci/jenkins/pull/6865

Reply:
Could someone explain to me what are we fixing exactly? I think I understand the overall concept, but I don't quite follow what the goal is. Prohibiting the view from having any Javascript is a closed road without a visible construction site...

The reason for inlining the JS is basically to keep the page load stateless. Prior to the change, in addition to the excessive recursion in fetching history, 3 separate calls to do the history fetch were done, each causing a recursion on its own. With this implementation the page and history is loaded once and the results are embedded into the view without needless ping pong in a semi-stateless manner.

Reply:
You need to move all JS code into a separate file `history.js` in the package `hudson.tasks.junit.History.history` and include it here with something like:

```
<st:bind var="view" value="${it}" />
<st:adjunct includes="hudson.tasks.junit.History.history"/>
```

But since you do not use the variable `view` at all, I think the bind can be removed totally.

So it will end up in 

```
<st:adjunct includes="hudson.tasks.junit.History.history"/>
```

The variables you are setting with Jelly need to be placed into data-attributes of a div (existing or new invisible one), where they can be retrieved later in JS.

Example PR from Daniel: https://github.com/jenkinsci/jenkins/pull/5514

Reply:
Done.

Reply:
Seems that my answer was parallel to your answer. But I see that you fixed it already 🎉 

I think the overall goal is to enable CSP overall in Jenkins, that means that no JS will be invoked anymore in any `script` tag when enabled. The only valid source for JS should be separate JS files. (This is still a long way to go but we should not introduce new `script` tags if possible)